Implement anvil destorying torches whilst falling down

geNAZt opened the pull request:
Fix cactus not having vanilla behaviour when sand falls on top.

## Introduction
Added a first entity behaviour. DestroyWhilstFalling can destroy/drop blocks whilst touching them.

### Relevant issues
<!-- List relevant issues here -->
* Fixes #2895 

## Changes
### Behavioural changes
The server now drops all sort of torches when a falling anvil touches them. Also sand gets dropped when hit by a cactus.

Commit:
Implement anvil destorying torches whilst falling down
Fix cactus not having vanilla behaviour when sand falls on top. Fixes #2895

dktapps commented on src/pocketmine/block/Cactus.php:
```diff
@@ -127,4 +129,18 @@ public function place(Item $item, Block $blockReplace, Block $blockClicked, int
+			return $parentDecision;
+		}
+
+		if ($block instanceof Sand) {
```
what about concrete?

geNAZt replied:
is there a list of blocks which has the same behaviour? does it work for all falling blocks?

dktapps replied:
As far as I know, yes.

dktapps commented on src/pocketmine/block/Block.php:
```diff
@@ -837,4 +839,18 @@ public function removeMetadata(string $metadataKey, Plugin $owningPlugin) : void
+	 * 			be destroyed and falling should continue
+	 */
+	public function canBlockLand(Fallable $block) : int{
+		return ($this->getId() > 0 and !$this->isTransparent() and !$this->canBeReplaced()) ?
```
We're moving away from legacy blockIDs, so please avoid this usage.

dktapps commented on src/pocketmine/entity/behaviour/Behaviour.php:
```diff
@@ -0,0 +1,38 @@
+	/** @var Entity */
+	protected $entity;
+
+	protected function __construct(Entity $entity) {
```
formatting issue here.

dktapps commented on src/pocketmine/entity/object/FallingBlock.php:
```diff
@@ -70,6 +81,10 @@ protected function initEntity(CompoundTag $nbt) : void{
 
 		$this->block = BlockFactory::get($blockId, $damage);
+		if ($this->block instanceof Anvil) {
+			$this->behaviours = [new DestroyWhileFalling($this,	[Block::TORCH, Block::COLORED_TORCH_RG, Block::COLORED_TORCH_BP,
```
Not a fan of this. This should be a property of the blocks themselves.

dktapps replied:
There's also a rogue tab in the argument list here.

geNAZt replied:
There is no property system for blocks in PMMP so i would have to generate a getter for it. Also this would not allow for any dynamic lists (since the behaviour is then dependent on the block property) of destoryable blocks.

dktapps reviewed (changes requested):
I've given this a quick look, but not in depth.

Commit:
In the test i made all falling block types do drop (even anvils do)

Commit:
Remove the tab
Remove the > 0 check. If a falling entity hits a block its solid. Air is
non solid so it always fires with a block other than 0

lukeeey commented:
I see you have given up on GoMint

dktapps commented on src/pocketmine/block/Block.php:
```diff
@@ -837,4 +839,18 @@ public function removeMetadata(string $metadataKey, Plugin $owningPlugin) : void
+	 * 			be destroyed and falling should continue
+	 */
+	public function canBlockLand(Fallable $block) : int{
+		return (!$this->isTransparent() and !$this->canBeReplaced()) ?
```
this will cause unintended side effects when dropped on glass.

dktapps commented on src/pocketmine/block/Cactus.php:
```diff
@@ -127,4 +129,14 @@ public function place(Item $item, Block $blockReplace, Block $blockClicked, int
+			return $parentDecision;
+		}
+
+		return FallingBlock::STATE_DROP_ITEM;
```
The logic in this function seems redundant. It seems like this could be reduced to a single line.

dktapps replied:
This problem has not been resolved.

Commit:
Remove conditions. All blocks which are solid enough to be counted as onGround condition fulfill the same conditions as the old checks

Commit:
Fix small formatting error

dktapps commented on src/pocketmine/entity/object/FallingBlock.php:
```diff
@@ -70,6 +81,10 @@ protected function initEntity(CompoundTag $nbt) : void{
 		$this->block = BlockFactory::get($blockId, $damage);
+		if ($this->block instanceof Anvil) {
+			$this->behaviours = [new DestroyWhileFalling($this, [Block::TORCH, Block::COLORED_TORCH_RG, Block::COLORED_TORCH_BP,
+					Block::LIT_REDSTONE_TORCH, Block::UNLIT_REDSTONE_TORCH, Block::UNDERWATER_TORCH], true)];
```
I don't like this. I'm trying to get rid of legacy block IDs. In addition, you're creating a whole new system here for one thing, which makes no sense and is inconsistent with the rest of the code.

It seems like this logic can be dropped and just add `canBlockLand()` implementations for these blocks.

dktapps replied:
In fact, the logic for this already exists, so this code can be removed entirely.

dktapps commented on src/pocketmine/entity/object/FallingBlock.php:
```diff
@@ -105,20 +120,42 @@ protected function entityBaseTick(int $tickDiff = 1) : bool{
+
+					case self::STATE_DESTROY_BLOCK:
+						$this->level->setBlock($blockDown, Block::get(Block::AIR));
+						$this->getLevel()->dropItem($blockDown, $blockDown->asItem());
```
This should use `Level->useBreakOn()`.

dktapps commented:
Closing due to lack of activity on requested changes.
